GLGUI: chat - added comments and minor runtime optimizations

0-8-15 opened the pull request:
Additionally I had to re-format the source according to widespread
conventions for readbility and used keyword like `when` and `unless`
to make the source more compact and easier to digest.

Sorry for the latter.

mgorges commented:
I have to disagree - there is not a single place in the entire codebase, where `unless` or `when` are used, as such this might confuse instead of help. You said not to make personal style related changed in one of your last pull request, but to me this clearly is, as this is your preference, not what the project has as a whole.

mgorges commented:
To make sure this is documented. The expression equivalence for `when` is:

``` 
(when <test>
  <expression1>
  <expression2>
)
```
is equivalent to
```
(if <test>
  (begin
    <expression1>
    <expression2> 
  )
)
```

mgorges commented:
The other case for `unless` includes an additional `not`, so:
```
(unless <test>
  <expression1>
  <expression2>
)
```
is equivalent to 
```
(if (not <test>)
  (begin
    <expression1>
    <expression2> 
  )
) 
```

mgorges commented on modules/ln_glgui/chat.scm:
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+                 (clipped (max valmin (min valmax calculated))))
+            (glgui-widget-set! g wgt 'offset clipped)
+            (glgui-widget-set! g wgt 'oldy my-fixed)
+            (unless drag (glgui-widget-set! g wgt 'drag #t)))))
```
This piece is new, it tests if `drag` is already set.

mgorges commented on modules/ln_glgui/chat.scm:
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+      (cond
+       ((not inside) ;; cancel drag
+        (if old (glgui-widget-set! g wgt 'old #f))
+        (if drag (glgui-widget-set! g wgt 'drag #f)))
```
this piece is new too.

mgorges commented on modules/ln_glgui/chat.scm:
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+              (when (and (>= cur 0) (>= cur val) (< cur n))
+                (glgui-widget-set! g wgt 'current cur)
+                (if (procedure? cb) (cb g wgt type mx my)))))
+        (if old (glgui-widget-set! g wgt 'old #f))
```
new test added

mgorges commented on modules/ln_glgui/chat.scm:
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+        (if old (glgui-widget-set! g wgt 'old #f))
+        (if drag (glgui-widget-set! g wgt 'drag #f)))))
+     (else ;; neither button1 nor motion
+      (unless inside
```
this piece is also new

mgorges commented on modules/ln_glgui/chat.scm:
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 (define (glgui:chat-input g wgt type mx my)
-  (let* ((x (fix (glgui-widget-get-dyn g wgt 'x)))
+  (let* ((mx-fixed (fix mx)) ;; Q: Can `mx` or `my` ever not be fixnums?
```
I don't know, but we use `fx` math later, which segfaults if it is not without pointing to where the error is, which is extremely annoying to debug.

0-8-15 commented:
Am Sat, 26 Sep 2020 13:02:00 -0700
schrieb Matthias Görges <notifications@github.com>:

> I have to disagree - there is not a single place in the entire
> codebase, where `unless` or `when` are used, as such this might
> confuse instead of help. You said not to make personal style related
> changed in one of your last pull request, but to me this clearly is,
> as this is your preference, not what the project has as a whole.

Sorry for that.  --  As I wrote in the commit message.

Initially I just wanted to understand the code (since I was hunting for
the bug I made elsewhere).

Then (with the background of ~30 years of Scheme experience) I felt that
the unusual formatting was the part holding me back.  There is a rather
long tradition to use automatic code formatting among most Schemers.
Essentially the editors do this for you.  A few more rules are usually
recommended, e.g., multiple let-bindings on a line are OK if the column
76 is not crossed, but for let* and letrec this is never done; always
have a space between adjacent closing and opening parenthesis (since
some reader extensions like SRFI-110 or gambits 'six' might convert
these things); nothing behind `begin` if the block spans multiple lines.

Additionally working with `git` (which I don't really like for a
version control system, but it won the world) taught me the habit to
let the editor automatically remove trailing white space.

Then I ran into the repeated use of `length` for the same input.
`length` is O(n) and thus usually avoided for performance.  As we are
in the many-time-per-second part of the code I just removed that one.
Same for the repeated use of `fix` on the same input.  As I now was in
the optimized-a-few-thing-here-mode I added those "don't modify a hash
table for no reason" checks.  For consistency I added another case in
the else branch.

At that point I was convinced that I should now understand the code and
can't find a bug there.  Hence I looked an found it elsewhere.

What to do with the change now?  Throw away the improvements just
because they contain what I would call a code cleanup because that
would include personal preferences?  Or propose it as a PR and have you
guy decide?

I opted for the latter.
